primitives/api/proc-macro/src/runtime_metadata.rs

```diff
@@ -0,0 +1,253 @@
+	let mut methods = Vec::new();
+
+	// Ensure that any function parameter that relies on the `BlockT` bounds
+	// also has `TypeInfo + 'static` bounds (required by `scale_info::meta_type`).
+	//
+	// For example, if a runtime API defines a method that has an input:
+	// `fn func(input: <Block as BlockT>::Header)`
+	// then the runtime metadata will imply `<Block as BlockT>::Header: TypeInfo + 'static`.
+	//
+	// This restricts the bounds at the metadata level, without needing to modify the `BlockT`
+	// itself, since the concrete implementations are already satisfying `TypeInfo`.
+	let mut where_clause = Vec::new();
+	for item in &decl.items {
+		// Collect metadata for methods only.
+		let syn::TraitItem::Method(method) = item else {
+			continue
+		};
+
+		// Collect metadata only for the latest methods.
+		let is_changed_in =
+			method.attrs.iter().any(|attr| attr.path.is_ident(CHANGED_IN_ATTRIBUTE));
+		if is_changed_in {
+			continue
+		}
+
+		let mut inputs = Vec::new();
+		let signature = &method.sig;
+		for input in &signature.inputs {
+			// Exclude `self` from metadata collection.
+			let syn::FnArg::Typed(typed) = input else {
+				continue
+			};
+
+			let pat = &typed.pat;
+			let name = format!("{}", quote!(#pat));
+			let ty = &typed.ty;
+			collect_where_bounds(ty).map(|ty_elem| where_clause.push(ty_elem));
+
+			inputs.push(quote!(
+				#crate_::metadata_ir::RuntimeApiMethodParamMetadataIR {
+					name: #name,
+					ty: #crate_::scale_info::meta_type::<#ty>(),
+				}
+			));
+		}
+
+		let output = match &signature.output {
+			syn::ReturnType::Default => quote!(#crate_::scale_info::meta_type::<()>()),
+			syn::ReturnType::Type(_, ty) => {
+				collect_where_bounds(ty).map(|ty_elem| where_clause.push(ty_elem));
+				quote!(#crate_::scale_info::meta_type::<#ty>())
+			},
+		};
+
+		// String method name including quotes for constructing `v15::RuntimeApiMethodMetadata`.
+		let method_name = format!("{}", signature.ident);
+		let docs = collect_docs(&method.attrs, &crate_);
+
+		// Include the method metadata only if its `cfg` features are enabled.
+		let attrs = filter_cfg_attributes(&method.attrs);
+		methods.push(quote!(
+			#( #attrs )*
+			#crate_::metadata_ir::RuntimeApiMethodMetadataIR {
+				name: #method_name,
+				inputs: #crate_::vec![ #( #inputs, )* ],
+				output: #output,
+				docs: #docs,
+			}
+		));
+	}
+
+	let trait_name_ident = &decl.ident;
+	let trait_name = format!("{}", trait_name_ident);
+	let docs = collect_docs(&decl.attrs, &crate_);
+	let attrs = filter_cfg_attributes(&decl.attrs);
+	// The trait generics where already extended with `Block: BlockT`.
+	let mut generics = decl.generics.clone();
+	for generic_param in generics.params.iter_mut() {
+		let syn::GenericParam::Type(ty) = generic_param else {
+			continue
+		};
+
+		// `scale_info::meta_type` requires `T: ?Sized + TypeInfo + 'static` bounds.
+		ty.bounds.push(parse_quote!(#crate_::scale_info::TypeInfo));
+		ty.bounds.push(parse_quote!('static));
```

Comment:
```suggestion
```

We should not need this? As below we would add the required bounds all the actual types being used.

Reply:
Have removed these extra bounds. Indeed they are not needed after extending the `get_type_param` to all param types not just the ones that have `Block: BlockT`. Thanks!